install_microshift.py

```diff
@@ -0,0 +1,178 @@
+import subprocess
+import time
+import host
+from logger import logger
+import os
+
+def cleanup_microshift(h: host.Host):
+    files = ["rhocp-4.14.toml", "fast-datapath.toml", "kickstart.ks"]
+
+    for file in files:
+        if os.path.exists(f"./{file}"):
+            os.remove(f"./{file}")
+
+    cmd = f"podman rm -f minimal-microshift-server"
+    output = h.run(cmd)
+    logger.info(output.out)
+
+    old_compose = h.run("composer-cli compose status").out
+    old_compose_splitlines = old_compose.splitlines()
+
+    for line in old_compose_splitlines[1:]:
+        compose_id = line.split(" ")[0] 
+        if 'FAILED' or 'FINISHED' not in line:
+            cmd = f" composer-cli compose cancel {compose_id}"
+            output = h.run(cmd)
+            logger.info(output.out)
+    
+        cmd = f"composer-cli compose delete {compose_id}"
+        output = h.run(cmd)
+        logger.info(output.out)
+
+def generate_kickstart(): 
+    insertion_point = 'cat << EOF > /etc/crio/openshift-pull-secret'
+
+    with open('pull_secret.json', 'r') as f_in:
+        file_contents = f_in.read()
+
+    with open('osbuild.ks', 'r') as f:
+        lines = f.readlines()
+
+    with open('kickstart.ks', 'w') as f_out:
+        for line in lines:
+            f_out.write(line)
+            if line.strip() == insertion_point:
+                f_out.write(file_contents)
+
+
+def microshift_iso_builder(h: host.Host):
```

Comment:
This function is too long. One of the reasons shorter functions are preferred is that it's easier to read. A long function is red flag and should raise your attention to want to refactor it. It's more nuanced than that, but this function is really long. The function should be written in a way that it breaks down into steps that are at a similar level of abstraction.  The first two functions are great: `cleanup_microshift` and `generate_kickstart`. Then, the function is written with very low level commands: `dnf install ...` and `systemctl`

If you read the code and tell yourself "here I'm preparing dependencies", then create a function `prep_dependencies` that does just that. Then you are adding files with compose-cli. You are doing:

```
    cmd = f"sudo composer-cli sources add fast-datapath.toml"
    output = h.run(cmd)
    logger.info(output.out)

    cmd = f"sudo composer-cli blueprints push minimal-microshift.toml"
    output = h.run(cmd)
    logger.info(output.out)

    cmd = f"sudo composer-cli blueprints depsolve minimal-microshift | grep microshift"
    output = h.run(cmd)
    logger.info(output.out)
```
I bet you copy pasted and changed part of the cmd every time. the fact that you copy pasted is again a red flag. This should prompt you to want to rewrite it as:

```
def composer_cli(something, something_else):
   ..
```

and then use that. That will immediately replace each set of 3 lines by 1 line and make the function shorter. That's one way to see if you should rewrite the code.

Continuing in the code, you ahve:
```
    while True:
        status_output = h.run(f"sudo composer-cli compose status").out

        build_status_line = next((line for line in status_output.split('\n') if build_id_two in line), None)

        if 'FINISHED' in build_status_line:
            break

        logger.info("Waiting on build to finish")
        time.sleep(2)
```

The log reveals immediately what this is doing: `Waiting on build to finish`.
Let's use a function yet again called "wait_for_build_to_finish`. 

Less things should be hardcoded too. You are using `final.iso` in this function, and then also somewhere else. That makes the code brittle. If you introduce a variable that you pass around, it will be more easy to maintain the code.

Let's start with the above changes, and I will review it again in more depth.

Reply:
This sounds good! Will make sure to refactor and make it shorter! Appreciate the detailed comment!